### PR TITLE
Add combined A/E fit routine

### PR DESCRIPTION
**Pull request by fhagemann**
Right now, this routine still uses the result of the old routine as starting parameters for the combined fit.

How to use it:
```julia
# ...

compton_bands  = aoe_config_ch.compton_bands
compton_window = aoe_config_ch.compton_window

result_fit, report_fit, compton_band_peakhists = nothing, nothing, nothing
try
    # get compton band peak histograms with generated peakstats
    compton_band_peakhists = generate_aoe_compton_bands(aoe, e_cal, compton_bands, compton_window)

    result_fit, report_fit = fit_aoe_compton(compton_band_peakhists.peakhists, compton_band_peakhists.peakstats, compton_bands,; uncertainty=true)
catch e
    @error "AoE compton bands cannot be fitted: $e"
    throw(ErrorException("AoE compton bands cannot be fitted"))
end

_compton_bands = [band for band in keys(result_fit) if result_fit[band].gof.pvalue >= p_value_cut] # end here 
μ = [result_fit[band].μ for band in _compton_bands]
σ = [result_fit[band].σ for band in _compton_bands]

# fit μ and σ with correction functions
result_corrections, report_corrections = nothing, nothing
try
    @time result_corrections, report_corrections = fit_aoe_corrections(_compton_bands, μ, σ,; e_expression = e_type_aoe_cal)
catch e
    @error "AoE corrections cannot be fitted: $e"
    throw(ErrorException("AoE corrections cannot be fitted"))
end

# use the result_corrections as input to the combined fit
combined_result = fit_aoe_compton_combined(compton_band_peakhists.peakhists, compton_band_peakhists.peakstats, compton_bands, result_corrections)
```

### CODE REVIEW

**Comment by fhagemann** [comment]
I was assigned to this PR? 
Does that mean I should merge it?

File: src/aoefit.jl [hunk 1]
```diff
@@ -327,3 +327,128 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+            # report[band] = report_band
+        end
+        return neg_log_likelihood
+    end
+    
+    # MLE
+    opt_r = optimize(f_loglike ∘ inverse(f_trafo), f_trafo(v_init))
+
+    v_ml = inverse(f_trafo)(Optim.minimizer(opt_r))
+
```
Comment (theHenks) [inline src/aoefit.jl:451]: Uncertainty handling by ForwardDiff similar to other firts

Comment (theHenks) [reply src/aoefit.jl:451]: The `uncertainty` should be a `kwarg` and calculated if
``` julia
if uncertainty && converged
...
```

File: src/aoefit.jl [hunk 2]
```diff
@@ -327,3 +327,128 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+            # save results
+            # result[band] = result_band
+            # report[band] = report_band
+        end
+        return neg_log_likelihood
+    end
+    
+    # MLE
+    opt_r = optimize(f_loglike ∘ inverse(f_trafo), f_trafo(v_init))
+
```
Comment (theHenks) [inline src/aoefit.jl:449]: Check if fit converged with `Optim.converged`

File: src/aoefit.jl [hunk 3]
```diff
@@ -327,3 +327,128 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+            end
+            # save results
+            # result[band] = result_band
+            # report[band] = report_band
+        end
+        return neg_log_likelihood
+    end
+    
+    # MLE
+    opt_r = optimize(f_loglike ∘ inverse(f_trafo), f_trafo(v_init))
```
Comment (theHenks) [inline src/aoefit.jl:448]: - Set `Optim.options` for number of iterations and time limit
- Check if different algorhithm should be used such as `LBFGS` maybe with `autodiff = :forward`

File: src/aoefit.jl [hunk 4]
```diff
@@ -327,3 +327,128 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+    # start values for MLE
+    v_init = mean(pseudo_prior)
+
+    # create loglikehood function
+    f_loglike = pars -> begin
+        
+        neg_log_likelihood = 0.0
+
+        # iterate throuh all peaks
+        for (i, band) in enumerate(compton_bands)
```
Comment (theHenks) [inline src/aoefit.jl:424]: Multithread single compton band fits with `Threads.@threads`

**Comment by theHenks** [comment]
Check binning in `generate_aoe_compton_bands` and adapt if necessary

File: src/aoefit.jl [hunk 5]
```diff
@@ -327,3 +327,183 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+            end
+        end
+        return sum(neg_log_likelihoods)
+    end
+    
+    # MLE
+    opt_r = optimize(f_loglike ∘ inverse(f_trafo), f_trafo(v_init), NelderMead(), Optim.Options(time_limit = 60, show_trace=false, iterations = 1000))
+
+    converged = Optim.converged(opt_r)
+    !converged && @warn "Fit did not converge"
```
Comment (fhagemann) [inline src/aoefit.jl:448]: Added convergence check

File: src/aoefit.jl [hunk 6]
```diff
@@ -327,3 +327,183 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+            param_covariance = zeros(length(v_ml), length(v_ml))
+        else
+            # Calculate the parameter covariance matrix
+            param_covariance = inv(H)
+        end
+        if ~isposdef(param_covariance)
+            param_covariance = nearestSPD(param_covariance)
+        end
+        # Extract the parameter uncertainties
+        v_ml_err = array_to_tuple(sqrt.(abs.(diag(param_covariance))), v_ml)
```
Comment (fhagemann) [inline src/aoefit.jl:471]: Added uncertainty handling

File: src/aoefit.jl [hunk 7]
```diff
@@ -327,3 +327,183 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+    # start values for MLE
+    v_init = mean(pseudo_prior)
+
+    # create loglikehood function
+    f_loglike = pars -> begin
+        
+        neg_log_likelihoods = zeros(typeof(pars.μA), length(compton_bands))
+
+        # iterate throuh all peaks (multithreaded)
+        Threads.@threads for i in eachindex(compton_bands)
```
Comment (fhagemann) [inline src/aoefit.jl:423]: Implemented multithreading

File: src/aoefit.jl [hunk 8]
```diff
@@ -327,3 +327,183 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+            catch e
+                @warn "Error fitting band $(compton_bands[i]): $e"
+                continue
+            end
+        end
+        return sum(neg_log_likelihoods)
+    end
+    
+    # MLE
+    opt_r = optimize(f_loglike ∘ inverse(f_trafo), f_trafo(v_init), NelderMead(), Optim.Options(time_limit = 60, show_trace=false, iterations = 1000))
```
Comment (fhagemann) [inline src/aoefit.jl:445]: Changed the number of iterations (still needs to be properly tested, these are arbitrary numbers)

File: src/aoefit.jl [hunk 9]
```diff
@@ -327,3 +327,183 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+            param_covariance = nearestSPD(param_covariance)
+        end
+        # Extract the parameter uncertainties
+        v_ml_err = array_to_tuple(sqrt.(abs.(diag(param_covariance))), v_ml)
+
+        # TODO: correctly combine p-values and residuals of the individual fits to the combined fit result
+        # get p-value 
+        # pval, chi2, dof = p_value(f_aoe_compton, h, v_ml)
+        # calculate normalized residuals
+        # residuals, residuals_norm, _, bin_centers = get_residuals(f_aoe_compton, h, v_ml)
```
Comment (fhagemann) [inline src/aoefit.jl:477]: `p-values, chi2, dof, residuals` still need to be implemented, probably by being passed from the individual fits and combined in the combined fit

File: src/aoefit.jl [hunk 10]
```diff
@@ -327,3 +327,183 @@ function fit_single_aoe_compton(h::Histogram, ps::NamedTuple; uncertainty::Bool=
...
+
+    # report = (
+    #         v = v_ml,
+    #         h = h,
+    #         f_fit = x -> Base.Fix2(f_aoe_compton, v_ml)(x),
+    #         f_sig = x -> Base.Fix2(f_aoe_sig, v_ml)(x),
+    #         f_bck = x -> Base.Fix2(f_aoe_bkg, v_ml)(x)
+    #     )
+
+    return result #, report
```
Comment (fhagemann) [inline src/aoefit.jl:507]: We still need to define the format of `report` and maybe implemented/adjust a plot recipe for the fit report.

File: src/aoefit.jl (hunk 3 above)
Comment (verenaaur) [reply src/aoefit.jl]: I changed the line to "opt_r = optimize(f_loglike ∘ inverse(f_trafo), f_trafo(v_init), LBFGS(); autodiff=:forward)" but now it is taking forever to run the "combined_result = LegendSpecFits.fit_aoe_compton_combined(compton_band_peakhists.peakhists, compton_band_peakhists.peakstats, compton_bands, result_corrections)" line.

Comment (fhagemann) [reply src/aoefit.jl]: Well, good to hear that you managed to get this running at all, but if it takes forever, let's leave it with `NelderMead` for now

**Comment by fhagemann** [comment]
`fit_aoe_compton_combined` now returns a `report` (format/naming can still be discussed/changed), which includes the fits and fit parameters to the individual Compton bands.
CC @theHenks @verenaaur

**Review by theHenks — APPROVED** [review]

File: src/aoefit.jl (hunk 9 above)
Comment (fhagemann) [reply src/aoefit.jl]: --> a39fa4d3eae3891f81919ce4ab0d1ad54df13009

**Comment by fhagemann** [comment]
So, to incorporate this code into the legend-julia-dataflow, one would need to insert
```julia
result_fit_combined, report_fit_combined = nothing, nothing
try
    result_fit_combined, report_fit_combined = fit_aoe_compton_combined(compton_band_peakhists.peakhists, compton_band_peakhists.peakstats, compton_bands, result_corrections; e_expression = e_type_aoe_cal, uncertainty = true);
catch e
    @error "AoE compton bands cannot be fitted using a combined fit: $e"
    throw(ErrorException("AoE compton bands cannot be fitted using a combined fit"))
end
```
after the block where `fit_aoe_corrections` is called.

The plots afterwards can be adjusted to also include the combined fit result by adding an extra argument:
```julia
# instead of: p = plot(report_correction.report_µ)
p = plot(report_correction.report_μ, report_fit_combined.report_μ)
```

One thing to note: The combined fits took way longer yesterday than they used to do before (which is surprising to me because I didn't touch any of the maximum likelihood fit code in the last commits I added (?) ). Not sure why this is and I hope this does not affect data production..

**Comment by fhagemann** [comment]
Never mind, false alarm: I reused a variable name and caused some type instabilities. This should be fixed with the latest force push.

**Comment by fhagemann** [comment]
There might still be changes on some internal functions (but just code style, no change of functionality),
so @theHenks you can safely merge this PR and use `fit_aoe_compton_combined` and the new plot recipe taking two arguments to start data production ;)

**Comment by fhagemann** [comment]
My last commit is just a clean-up, it does not change any functionality.
It just increases type stability.

**Review by theHenks — APPROVED** [review]
Very nice! :)

**Pull request merged**